### ISSUE TITLE
[RFC] Support /proc/id/statm file 

### ISSUE THREAD

**Issue by 3AHAT0P** [occlum/occlum#1542]
- Feature Name: Support /proc/id/statm file 
- Start Date: 2024-05-17

# Summary
[summary]: #summary

Support standard Linux /proc/id/statm file with information about used memory by process

# Motivation
[motivation]: #motivation

Many software use information about process (self) used RAM for manage threads and children process. Main of possibility getting usage mem info is read data from /proc/self/statm. That file is the standard of Linux system.
For example in the python script we can use `psutil.Process().memory_info().rss` but in Occlum LibOS environment it raise the exception about "statm not found".

Currently I known at least one software - Catboost (similar a xgboost) which needed information from statm file.

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

       /proc/pid/statm
              Provides information about memory usage, measured in
              pages.  The columns are:

                  size       (1) total program size (same as VmSize in /proc/pid/status)
                  resident   (2) resident set size (inaccurate; same as VmRSS in /proc/pid/status)
                  shared     (3) number of resident shared pages (i.e., backed by a file) (inaccurate; same as RssFile+RssShmem in /proc/pid/status)
                  text       (4) text (code)
                  lib        (5) library (unused since Linux 2.6; always 0)
                  data       (6) data + stack
                  dt         (7) dirty pages (unused since Linux 2.6; always 0)

              Some of these values are inaccurate because of a kernel-
              internal scalability optimization.  If accurate values are
              required, use /proc/pid/smaps or /proc/pid/smaps_rollup
              instead, which are much slower but provide accurate,
              detailed information.

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

-

# Drawbacks
[drawbacks]: #drawbacks

-

# Rationale and alternatives
[rationale-and-alternatives]: #rationale-and-alternatives

As alternative user could use syscall getrusage with RUSAGE_SELF as "who". But in this case we couldn't get total memory size, instead of RSS we just get maxrss. We can get this information only about current process, and don't about other process. Is not the same, but can use in some case.

Also, user could read /proc/id/maps and summarize info. In this case user could take info about total size, but not RSS (I don't sure about that a lot). This variant very hard for user code implementation, but in some cases goodest options then getrusage syscall.

# Prior art
[prior-art]: #prior-art

-

# Unresolved questions
[unresolved-questions]: #unresolved-questions

-

# Future possibilities
[future-possibilities]: #future-possibilities

-

**Comment by jessehui** [comment]
@3AHAT0P Thank you for using Occlum. We will take a look at this feature. Could you provide us with a simple demo ? A PR with demo would be even better. It will assist us in understanding the practical use case and evaluating its correctness.

**Comment by 3AHAT0P** [comment]
@jessehui If I understand correct you want to see PR with new demo into demos folder which used /proc/pid/statm (like a hello_c or etc)? 
If is it I will do it soon.

**Comment by jessehui** [comment]
Yes. But a hello_c demo to just `cat /proc/id/statm` is not enough. I assume a real-world demo can be provided like `Catboost ` you mentioned.

**Comment by jessehui** [comment]
@3AHAT0P Thank you for your contribution! I have tested the demo and it behaves exactly like you said. 

However, after I added support for the `/proc/pid/statm`, the log shows it also need `getrusage` syscall. So I added a dummy implementation of the `getrusage` to give all zero and it can run. You can use my PR here: #1544 

But I don't think it works correctly because in the end, it throws `139` errno and it is due to the `SIGSEGV` signal.

![截屏2024-05-22 20 04 43](https://github.com/occlum/occlum/assets/13930406/94fd32ed-8deb-4987-98fd-56c505ac6c1e)

So I did a little digging, and it looks like the application visit a memory range with `PROT_NONE` set and that's why `SIGSEGV` is sent.
```log
[T88][#48476][HandleException][vm] enclave page fault caught, pf_addr = 0x7f9d5b400000, error code = 32775[0m
[T88][#48476][HandleException][vm] PF vma = VMArea { range: VMRange { start: 0x7f9d5b400000, end: 0x7f9d68000000, size: 0xcc00000 }, perms: NONE, file_backed: None, access: Private(2)}}
```

Could you provide a simpler catboost demo with less data and shorter python code for me to debug ?

**Comment by 3AHAT0P** [comment]
@jessehui Wow! Thank you for your fast answer!
I will try your PR locally, investigate into Catboost code and return back with new details for you.

**Comment by 3AHAT0P** [comment]
@jessehui I found strange thing.
I have modified demo of hello_c with next code 
```clang
#include <sys/time.h>
#include <sys/resource.h>
#include <unistd.h>
#include <stdio.h>

int main()
{
    struct rusage usage;
    struct rusage usage2;
    struct timeval start, end;
    int i, j, k, g = 0;

    getrusage(RUSAGE_SELF, &usage);
    start = usage.ru_stime;
    for (g = 0; g < 3; g++)
    {
        for (i = 0; i < 10000; i++)
        {
            /* Double loop for more interesting results. */
            for (j = 0; j < 10000; j++)
            {
                k += 20;
            }
        }

        printf("End of global loop %ld times\n", g);
    }
    getrusage(RUSAGE_SELF, &usage2);
    end = usage2.ru_stime;

    printf("Started at: %ld.%lds\n", start.tv_sec, start.tv_usec);
    printf("Ended at: %ld.%lds\n", end.tv_sec, end.tv_usec);
    printf("ru_maxrss at: %ld %ld\n", usage.ru_maxrss, usage2.ru_maxrss);
    printf("ru_ixrss at: %ld %ld\n", usage.ru_ixrss, usage2.ru_ixrss);
    printf("ru_idrss at: %ld %ld\n", usage.ru_idrss, usage2.ru_idrss);
    printf("ru_isrss at: %ld %ld\n", usage.ru_isrss, usage2.ru_isrss);
    printf("ru_minflt at: %ld %ld\n", usage.ru_minflt, usage2.ru_minflt);
    printf("ru_majflt at: %ld %ld\n", usage.ru_majflt, usage2.ru_majflt);
    printf("ru_nswap at: %ld %ld\n", usage.ru_nswap, usage2.ru_nswap);
    printf("ru_inblock at: %ld %ld\n", usage.ru_inblock, usage2.ru_inblock);
    printf("ru_oublock at: %ld %ld\n", usage.ru_oublock, usage2.ru_oublock);
    printf("ru_msgsnd at: %ld %ld\n", usage.ru_msgsnd, usage2.ru_msgsnd);
    printf("ru_msgrcv at: %ld %ld\n", usage.ru_msgrcv, usage2.ru_msgrcv);
    printf("ru_nsignals at: %ld %ld\n", usage.ru_nsignals, usage2.ru_nsignals);
    printf("ru_nvcsw at: %ld %ld\n", usage.ru_nvcsw, usage2.ru_nvcsw);
    printf("ru_nivcsw at: %ld %ld\n", usage.ru_nivcsw, usage2.ru_nivcsw);
    return 0;
}
```

And it's demo on `0.30.1-ubuntu20.04` print next stats 
```
End of global loop 0 times
End of global loop 1 times
End of global loop 2 times
Started at: 656.1144s
Ended at: 0.140394331373632s
ru_maxrss at: 144 56
ru_ixrss at: 24 8
ru_idrss at: 220 4096
ru_isrss at: 24 140394335567872
ru_minflt at: 1360 0
ru_majflt at: 2705 140394331375040
ru_nswap at: 0 0
ru_inblock at: 71 0
ru_oublock at: 0 0
ru_msgsnd at: 0 140394335708929
ru_msgrcv at: 0 1
ru_nsignals at: 0 140394335708521
ru_nvcsw at: 7 0
ru_nivcsw at: 140394338274048 0
```

But in your branch that demo print that
```
End of global loop 0 times
End of global loop 1 times
End of global loop 2 times
Started at: 0.0s
Ended at: 0.0s
ru_maxrss at: 0 0
ru_ixrss at: 0 0
ru_idrss at: 0 0
ru_isrss at: 0 0
ru_minflt at: 0 0
ru_majflt at: 0 0
ru_nswap at: 0 0
ru_inblock at: 0 0
ru_oublock at: 0 0
ru_msgsnd at: 0 0
ru_msgrcv at: 0 0
ru_nsignals at: 0 0
ru_nvcsw at: 0 0
ru_nivcsw at: 0 0
```

Yes, I understand you add dummy implementation of the `getrusage` but I don't understand why it works fine in the `0.30.1-ubuntu20.04`. 
Or it's values work only in SIM mode and get's from host system?

**Comment by jessehui** [comment]
I think the reason is that you didn't check the return value of the `getrusage`.  And the `usage` structs are not initialized. And thus, just print some random values of the corresponding memory.

**Comment by 3AHAT0P** [comment]
Ouch. This is very good catch. Thanks a lot!
